src/Primitives/src/ChangeToken.cs

```diff
@@ -67,26 +50,91 @@ public static IDisposable OnChange<TState>(Func<IChangeToken> changeTokenProduce
                 throw new ArgumentNullException(nameof(changeTokenConsumer));
             }
 
-            Action<object> callback = null;
-            callback = (s) =>
+            return new ChangeTokenRegistration<TState>(changeTokenProducer, changeTokenConsumer, state);
+        }
+
+        private class ChangeTokenRegistration<TState> : IDisposable
+        {
+            private readonly Func<IChangeToken> _changeTokenProducer;
+            private readonly Action<TState> _changeTokenConsumer;
+            private readonly TState _state;
+            private object _disposable;
+
+            private static readonly object _disposedSentinel = new object();
+
+            public ChangeTokenRegistration(Func<IChangeToken> changeTokenProducer, Action<TState> changeTokenConsumer, TState state)
+            {
+                _changeTokenProducer = changeTokenProducer;
+                _changeTokenConsumer = changeTokenConsumer;
+                _state = state;
+
+                var token = changeTokenProducer();
+
+                RegisterChangeTokenCallback(token);
+            }
+
+            private void OnChangeTokenFired()
             {
                 // The order here is important. We need to take the token and then apply our changes BEFORE
                 // registering. This prevents us from possible having two change updates to process concurrently.
                 //
                 // If the token changes after we take the token, then we'll process the update immediately upon
                 // registering the callback.
-                var t = changeTokenProducer();
+                var token = _changeTokenProducer();
+
                 try
                 {
-                    changeTokenConsumer((TState)s);
+                    _changeTokenConsumer(_state);
                 }
-                finally // We always want to ensure the callback is registered
+                finally
                 {
-                    t.RegisterChangeCallback(callback, s);
+                    // We always want to ensure the callback is registered
+                    RegisterChangeTokenCallback(token);
                 }
-            };
+            }
+
+            private void RegisterChangeTokenCallback(IChangeToken token)
+            {
+                var registraton = token.RegisterChangeCallback(s => ((ChangeTokenRegistration<TState>)s).OnChangeTokenFired(), this);
+
+                SetDisposable(registraton);
```

Comment:
Are you concerned at all about the case of a bad `IChangeToken` that doesn't implicitly dispose registrations after it fires? I guess there's not much we could do to mitigate that, just thinking about it since we're only going dispose the 'current' registration.

Reply:
Yes, but that's another change, my previous attempt to dispose the previous registrations broke configuration so there's something else going on there, that needs to be looked into.